Doc/library/itertools.rst

```diff
@@ -435,8 +435,13 @@ loops that truncate the stream.
           # islice('ABCDEFG', 2, 4) --> C D
           # islice('ABCDEFG', 2, None) --> C D E F G
           # islice('ABCDEFG', 0, None, 2) --> A C E G
+          # it = iter('abcdefghi')
```

Comment:
Instead, please add a new class, TestPurePythonRoughEquivalents, with a test_islice_recipe method that runs all the test_islice() tests against the pure python version of islice().

While we're at it, modify the consume() recipe to have *n* default to *None*:

    - def consume(iterator):
    + def consume(iterator, n=None):

Let's also add a test to Lib/test/test_itertools.py libreftest section so that the *consume()* code gets exercised

    >>> it = iter(range(10))
    >>> consume(it, 3)
    >>> next(it)
    3
    >>> consume(it)
    >>> next(it, 'Done')
    'Done'

Reply:
Thank you for the suggestions.  I have a question about how far the pure python recipe should be changed to get the tests to work.  With adding the tests for the islice_recipe, I'm able to do this to avoid duplicating code:
```python
class TestPurePythonRoughEquivalents(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        global islice
        cls.orig_islice = islice
        islice = cls.islice

    @classmethod
    def tearDownClass(cls):
        global islice
        islice = cls.orig_islice

    @staticmethod
    def islice(iterable, *args):
        s = slice(*args)
        it = iter(range(s.start or 0, s.stop or sys.maxsize, s.step or 1))
        try:
            nexti = next(it)
        except StopIteration:
            # Consume *iterable* up to the *start* position.
            for i, element in zip(range(s.start or 0), iterable):
                pass
            return
        for i, element in enumerate(iterable):
            if i == nexti:
                yield element
                try:
                    nexti = next(it)
                except StopIteration:
                    return

    def test_islice_recipe(self):
        TestBasicOps().test_islice()
```

However, `test_islice()` in `TestBasicOps()` has tests that check for invalid arguments with TypeError and ValueError, which the current recipe doesn't have.  Should I expand the recipe for these cases?

Also, there's a test for issue 10323 that fails with the pure python version:
```python
        c = count()
        self.assertEqual(list(islice(c, 1, 3, 50)), [1])
        self.assertEqual(next(c), 3)
```
It returns 2 instead of 3.

And the test for pickling fails with `TypeError: can't pickle generator objects`.

I don't know the way to get around the pickle error, but for the others, should I modify the recipe to pass all the tests?

Thanks!

Reply:
We don't need to modify the recipe or try to pass all of the regular islice() tests.   Instead, we just want to test the recipe to make sure its basic functionality is okay.  The pure python rough equivalent is allowed to not be a class, not be pickleable, and to omit error checks, etc, but it does need to be good at slicing :-)  The problem we're fixing here arose because the islice(it, n, n) case wasn't tested to see if it advanced the iterator.  That is core functionality, so it warrants a test.  

Also remember, the main purpose of the pure python rough equivalents it to augment the docs to provide additional insights that aren't easily conveyed in English.  It needs to be short and expressive.  A true exact equivalent, drop-in substitute would likely lose the pithy explanatory value.

Reply:
Thank you for explaining that.

A few notes on what the changes I made.
1.  I changed the recipe for the stop point.  This had a test referencing bpo-10323 and, to your point, it affects the slicing.  But, it also makes the recipe less pithy, so it may be too much of an edge case to worry about in the recipe?  I wanted to highlight it in this commit even though you ultimately may not want to add it.
2.  There were no existing test cases where start = stop in test_islice, so I added one under the comment for consume.  I thought maybe it should replace the existing test case for consume since it's more of an edge case?
```python
        # Test number of items consumed     SF #1171417
        it = iter(range(10))
        self.assertEqual(list(islice(it, 3)), list(range(3)))
        self.assertEqual(list(it), list(range(3, 10)))

        it = iter(range(10))
        self.assertEqual(list(islice(it, 3, 3)), [])
        self.assertEqual(list(it), list(range(3, 10)))
```
I added the second test to reflect this issue.

Thanks!